src/ethereum/istanbul/vm/__init__.py

```diff
@@ -0,0 +1,85 @@
+"""
+Ethereum Virtual Machine (EVM)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. contents:: Table of Contents
+    :backlinks: none
+    :local:
+
+Introduction
+------------
+
+The abstract computer which runs the code stored in an
+`eth1spec.eth_types.Account`.
+"""
+
+from dataclasses import dataclass
+from typing import Dict, List, Optional, Set, Tuple, Union
+
+from ethereum.base_types import U256, Bytes, Bytes0, Uint
+from ethereum.crypto.hash import Hash32
+
+from ..eth_types import Address, Log
+from ..state import State
+
+__all__ = ("Environment", "Evm", "Message")
+
+
+@dataclass
+class Environment:
+    """
+    Items external to the virtual machine itself, provided by the environment.
+    """
+
+    caller: Address
+    block_hashes: List[Hash32]
+    origin: Address
+    coinbase: Address
+    number: Uint
+    gas_limit: Uint
+    gas_price: U256
+    time: U256
+    difficulty: Uint
+    state: State
+
+
+@dataclass
+class Message:
+    """
+    Items that are used by contract creation or message call.
+    """
+
+    caller: Address
+    target: Union[Bytes0, Address]
+    current_target: Address
+    gas: U256
+    value: U256
+    data: Bytes
+    code_address: Optional[Address]
+    code: Bytes
+    depth: Uint
+    should_transfer_value: bool
+    is_static: bool
+
+
+@dataclass
+class Evm:
+    """The internal state of the virtual machine."""
+
+    pc: Uint
+    stack: List[U256]
+    memory: bytearray
+    code: Bytes
+    gas_left: U256
+    env: Environment
+    valid_jump_destinations: Set[Uint]
+    logs: Tuple[Log, ...]
+    refund_counter: int
```

Comment:
On a per-frame bases, EIP-2200 allows the refund counter to be negative. U256 is no longer valid.

Reply:
Ah! That's weird, but makes sense.

Now, do we backport this? On one hand, the refund counter would never have gone negative, but there are a couple changes in this diff that aren't really meaningful because of it.

Reply:
Yes. We should back port this to other forks.

Reply:
Created Issue #509